### PR TITLE
[CSE-231] 'Address does not exist' error for existing addresses.

### PR DESCRIPTION
**Pull request by ksaric**
https://issues.serokell.io/issue/CSE-231

Uncomment the necessary lines in the test, run the test, and see how it fails.

Before:

`http://localhost:8100/api/addresses/summary/DdzFFzCqrhsqU7rLVY4rjYgmHidTDLLfNpLDQk85QwCBu8WQW8Snpmq4rEmeZ2ze96WzBrLEt8FjTWU4f7cG7bsymDrpfKpDC5Jknpda`

`{"Left":"Address length is not equal to 32, can't be redeeming pk"}`

After:

`http://localhost:8100/api/addresses/summary/DdzFFzCqrhsqU7rLVY4rjYgmHidTDLLfNpLDQk85QwCBu8WQW8Snpmq4rEmeZ2ze96WzBrLEt8FjTWU4f7cG7bsymDrpfKpDC5Jknpda`

`{"Right":{"caAddress":"DdzFFzCqrhsqU7rLVY4rjYgmHidTDLLfNpLDQk85QwCBu8WQW8Snpmq4rEmeZ2ze96WzBrLEt8FjTWU4f7cG7bsymDrpfKpDC5Jknpda","caType":"CPubKeyAddress","caTxNum":2,"caBalance":{"getCoin":"0"},"caTxList":[{"ctbId":"ac439284933ba4a2bc81e5e8995c4a0c0a39256884d905e74ffcdf7ba9b257e7","ctbTimeIssued":1506329611,"ctbInputs":[["DdzFFzCqrhsqU7rLVY4rjYgmHidTDLLfNpLDQk85QwCBu8WQW8Snpmq4rEmeZ2ze96WzBrLEt8FjTWU4f7cG7bsymDrpfKpDC5Jknpda",{"getCoin":"1000000"}]],"ctbOutputs":[["DdzFFzCqrhsgHx6CUkwaLwABsZLtFZpuQsTbXpMkinerzLPCX9aNdpTVvj4P7uEmytxD8wW34y1A3f7awj1vGQMU3dBjv56619WXLjCo",{"getCoin":"329106"}],["DdzFFzCqrhsuXWViKNnvsC5NdWhT4NEw5FeDtzdm9WbzzDTTTrFSjRAPhiPhE258dt7hfHmQAHo6PNEkfzXZ7ZrF4WAiLYqPCmgUrc5G",{"getCoin":"500000"}]],"ctbInputSum":{"getCoin":"1000000"},"ctbOutputSum":{"getCoin":"829106"}},{"ctbId":"dbc87d198fc9f6067c8e65f89da60a944e1ef52f0c539d980ebbc95f3dc71b65","ctbTimeIssued":1506329391,"ctbInputs":[["Ae2tdPwUPEZ8fbeSrt7u4SfYF3GDCMBmqX4KaKygHjjcm8AjJH4R9Nbpp6u",{"getCoin":"1000000"}]],"ctbOutputs":[["DdzFFzCqrhsqU7rLVY4rjYgmHidTDLLfNpLDQk85QwCBu8WQW8Snpmq4rEmeZ2ze96WzBrLEt8FjTWU4f7cG7bsymDrpfKpDC5Jknpda",{"getCoin":"1000000"}]],"ctbInputSum":{"getCoin":"1000000"},"ctbOutputSum":{"getCoin":"1000000"}}]}}`

Also you can run the mainnet and try it there.

### CODE REVIEW

File: explorer/src/Pos/Explorer/Web/Server.hs [hunk 1]
```diff
@@ -711,8 +715,31 @@ topsortTxsOrFail f =
...
+--             -- * cardano-sl/tools/src/addr-convert/Main.hs
+--             unless (BS.length addr == 32) $
+--                 throwM badAddressLength
+--             pure $ makeRedeemAddress $ redeemPkBuild addr
+--         Nothing ->
+--             -- cAddr is in Cardano address format
+--             either badCardanoAddress pure (fromCAddress cAddr)
+--   where
+--     badAddressLength = Internal "Address length is not equal to 32, can't be redeeming pk"
+--     badCardanoAddress = const $ throwM $ Internal "Invalid Cardano address!"
```
Comment (dcoutts) [inline explorer/src/Pos/Explorer/Web/Server.hs]: In particular, we're removing a bunch of checks here. What is the idea behind this?

Also, don't leave commented-out code like this.

File: explorer/src/Pos/Explorer/Web/Server.hs [hunk 2]
```diff
@@ -724,15 +751,16 @@ cAddrToAddr cAddr@(CAddress rawAddrText) =
...
         Nothing ->
             -- cAddr is in Cardano address format
-            either badCardanoAddress pure (fromCAddress cAddr)
+            error "Undecoded address"
+            -- either badCardanoAddress const (fromCAddress cAddr)
   where
-    badAddressLength = Internal "Address length is not equal to 32, can't be redeeming pk"
-    badCardanoAddress = const $ throwM $ Internal "Invalid Cardano address!"
+    badAddressLength  = error "Address length is not equal to 32, can't be redeeming pk"
+    -- badCardanoAddress = error "Invalid Cardano address!"
```
Comment (dcoutts) [inline explorer/src/Pos/Explorer/Web/Server.hs]: Why pure exceptions now? And this looks half-finished, things still commented out.

File: explorer/src/Pos/Explorer/Web/Server.hs [hunk 3]
```diff
@@ -711,8 +715,31 @@ topsortTxsOrFail f =
...
+--             pure $ makeRedeemAddress $ redeemPkBuild addr
+--         Nothing ->
+--             -- cAddr is in Cardano address format
+--             either badCardanoAddress pure (fromCAddress cAddr)
+--   where
+--     badAddressLength = Internal "Address length is not equal to 32, can't be redeeming pk"
+--     badCardanoAddress = const $ throwM $ Internal "Invalid Cardano address!"
+
+cAddrToAddr :: CAddress -> Address
+cAddrToAddr (CAddress rawAddrText) =
```
Comment (dcoutts) [inline explorer/src/Pos/Explorer/Web/Server.hs]: Why did you change the type of `cAddrToAddr` when it's not even used in this module now? Is it used elsewhere? If so it would need to be fixed, if not it does not need to be exported.

File: explorer/src/Pos/Explorer/Web/Server.hs (hunk 1 above)
Comment (ksaric) [reply explorer/src/Pos/Explorer/Web/Server.hs]: The checks were removed since only RSCoin addresses have that invariant, other addresses don't have it.
I commented out the code since I wanted to leave a function people can run in a test to see why it fails and left the original commented so they can see the difference.
It's not like I usually leave commented out code around.

File: explorer/src/Pos/Explorer/Web/Server.hs (hunk 3 above)
Comment (ksaric) [reply explorer/src/Pos/Explorer/Web/Server.hs]: It's being used in the test, there are comments in the PR for running the tests.

File: explorer/src/Pos/Explorer/Web/Server.hs (hunk 2 above)
Comment (ksaric) [reply explorer/src/Pos/Explorer/Web/Server.hs]: Pure exceptions so I can test them quickly. Commented out the things that are not interesting to us.

File: explorer/src/Pos/Explorer/Web/Server.hs (hunk 1 above)
Comment (ksaric) [reply explorer/src/Pos/Explorer/Web/Server.hs]: Comment from @gromakovsky:

Decoded RSCoin (AVVM) address must have exactly 32 bytes, but Cardano address is different and its length is not fixed. I don't know how `cAddrToAddr` is used, but it makes sense only for RSCoin addresses.

**Comment by dcoutts** [comment]
And the patch includes `node/mainnet-genesis.json` which is presumably a mistake.

**Comment by ksaric** [comment]
> And the patch includes `node/mainnet-genesis.json` which is presumably a mistake.

@dcoutts I don't know If it's supposed to be available or not :man_shrugging: 
There is a script for connecting both `wallet` and `explorer` to _mainnet_. The PR was created by @domenkozar (https://github.com/input-output-hk/cardano-sl/pull/1645) and I merged it here so someone reviewing this can confirm the fix.
Since there is no other way to confirm it really works, since _DEV_ mode no longer works, and the small test that is not convincing enough to me to approve something for _mainnet_.

**Comment by ksaric** [comment]
Let me sum up the information that I have, and present my interpretation why this fix fixes things.

The first error was reported to be - `{"Left":"Address length is not equal to 32, can't be redeeming pk"}`.

So the first thing I noticed was that the address length seems to be different:
- old - `Sfpj3GbcsazoxEFvidt6rfedaX6PiXnYpYXTfj8hEgXfUFzk1kPWCFEFrecC9iWs7QP7yktEih4YuygF1JitxKze4z3bUFs9J`
- new - `DdzFFzCqrht8wAQiwNCromuPxNjQoK2Cs2vMiVFwFYYAQCcA1nPs7BMXFYhZZVBYhAKexYhaiA8xCUW8EEnc4Wdn6X5zD7R9xcabHip8`

I asked @gromakovsky .
He replied that address serialization changed recently - https://github.com/input-output-hk/cardano-sl/pull/1543 and that the length of the `Address` is not fixed.

Then I asked @adinapoli-iohk about that change, and he said similar things - seems the address length is arbitrary.

The important comment is from @gromakovsky :
Decoded RSCoin (AVVM) address must have exactly 32 bytes, but Cardano address is different and its length is not fixed. I don't know how `cAddrToAddr` is used, but it makes sense only for RSCoin addresses.

I tried to test both addresses on the `cardano-sl-1.0` branch.
Seems the new format is valid and works. 
The old one returns `predicate failed on: Left "decodeFull failed for Address: DeserialiseFailure 1 \"decodeCrcProtected: Proxy * ((AbstractHash Blake2b_224 Address'),(Attributes AddrAttributes),AddrType) failed the size check. Expected 2, found 4\""`.
So the serialization/deserialization seems to be working correctly.

So, it must be the address size is now arbitrary while we check it to be 32 bytes.

Since I couldn't test on _DEV_ mode and waited for @domenkozar to create a script that can connect to _mainnet_. 
He completed the script soon and you can see the PR here - https://github.com/input-output-hk/cardano-sl/pull/1644.

Until he created these changes and that PR for connecting to _mainnet_, I wanted to test it locally. 
So I created changes in the original function `cAddrToAddr` to have pure exceptions because there were some complications while I attempted to test the code as is.
I changed the original function `cAddrToAddr` and left the old one so one doesn't mistake the new function for the culprit.

The test failed and it showed that function `cAddrToAddr` can't process the new addresses - the same error occurs - `{"Left":"Address length is not equal to 32, can't be redeeming pk"}`.

So when the PR for _mainnet_ was done, I run the new changes and the addresses worked.

**In conclusion, seems that the address format was changed to be arbitrarily sized, and the invariant `size == 32` no longer holds. That's why this PR fixes that address issue.**

Phew. :tired_face:

**Comment by akegalj** [comment]
@Martoon-00 
> So, it must be the address size is now arbitrary while we check it to be 32 bytes.
is this true? This doesn't sound good to me. I thought address size is fixed (can be shorter and longer - the long containing root key)

**Comment by Martoon-00** [comment]
Format of address has changed recently, now address payload is signed.
Anyway, expecting some fixed address size seems pointless

File: explorer/src/Pos/Explorer/Web/Server.hs [hunk 4]
```diff
@@ -367,7 +368,10 @@ getAddressSummary
     => CAddress
     -> m CAddressSummary
 getAddressSummary cAddr = do
-    addr <- cAddrToAddr cAddr
+    -- let addr = cAddrToAddr cAddr
```
Comment (dcoutts) [inline explorer/src/Pos/Explorer/Web/Server.hs]: Can drop this commented out code too.
